src/core/exceptions.c

```diff
@@ -346,13 +353,54 @@ static void run_handler(MVMThreadContext *tc, LocatedHandler lh, MVMObject *ex_o
         /* Deliberate fallthrough to unwind below. */
         MVM_FALLTHROUGH
 
-    case MVM_EX_ACTION_GOTO:
+    case MVM_EX_ACTION_GOTO: ;
+        MVMPostUnwind cch = NULL;
+        if (category == MVM_EX_CAT_RETURN) {
+            cch = &cleanup_control_handler;
+            MVMuint8 set_ach = 0;
+            if (tc->active_control_handler == NULL)
+                set_ach = 1;
+            else {
+                /* Search down the stack for a control handler. And see if ours or
+                 * the saved one is deeper in the stack. The deeper one wins. */
+                MVMCallStackIterator iter;
+                MVM_callstack_iter_frame_init(tc, &iter, tc->stack_top);
+                MVMuint8 frame_found = 0;
+                while (MVM_callstack_iter_move_next(tc, &iter)) {
+                    MVMFrame *cur_frame = MVM_callstack_iter_current_frame(tc, &iter);
+                    if (cur_frame == lh.frame) {
+                        lh.frame = tc->active_control_handler->frame;
+                        lh.handler = tc->active_control_handler->handler;
+                        lh.jit_handler = tc->active_control_handler->jit_handler;
+                        tc->last_payload = tc->active_control_handler->ex_obj;
+                        frame_found = 1;
+                        break;
+                    }
+                    else if (cur_frame == tc->active_control_handler->frame) {
+                        MVM_free(tc->active_control_handler);
```

Comment:
Does tc->active_control_handler->ex_obj need to be freed separately? Is this a memory leak?

Reply:
At least this question is an easy one: an `MVMObject *` is a pointer to an object managed by the GC, and those will be cleared up by the GC (in fact, to free it would be an error).